Jay ) 엑티비티 전환 및 앨범에서 사진 가져오기 기능 구현

shim506 opened the pull request:
1. constraintLayout 의 컴포넌트 들을 xml 코드로 생성했습니다.
2. ImageView에  이미지를 표시했습니다.
3. 프로젝트의 asset 폴더를 생성하고 이미지를 넣어 사용했습니다.
4. 앨범에서 이미지를 불러왔습니다.

Q) 원하는 뷰를 앞-뒤 순서로 배치한다는 말을 xml 에서 elevation 을 통해서 구현했다고 생각했는데 맞는지 의문입니다.  사진을 가져오고 그 후 코드를 통해서 순서를 바꾸는 것을 의도하신 건가요?

shim506 committed:
과제 5 진행중 임시저장( pr 처리를 위한 )

shim506 committed:
MyImage 클래스 생성및 기능 구현

shim506 committed:
FEAT : WrongPathException 클래스 추가

이미지의 잘못된 경로에 대한 예외를 생성

shim506 committed:
FEAT : 버튼 클릭시 사진 변화 구현

- 사진 로딩 하기 구현
- 경로검사 함수 구현

shim506 committed:
DOCS : 클릭시 사진 변경에 기능에 대한 리드미 정리

shim506 committed:
FEAT : 두번째 Activity 로의 이동 구현

- fab 버튼 생성
- 두번째 Activity  연결

shim506 committed:
FEAT : ViewBinding 적용

shim506 committed:
FEAT : 액자 화면 구성

- png 파일 받아서 배경으로 지정

shim506 committed:
FEAT : 겔러리 이미지 불러 오기 기능 구현

- 겔러리 이미지 가져오기
- Glide 사용

shim506 committed:
DOCS : 추가학습 사항 정리

- viewbinding 개념 정리
- glide 라이브러리

shim506 committed:
STYLE: 코드 포멧팅 수정 및 기타 작업

- 코트 포멧팅 수정
- 리드미 step1 부분 옮김 , 액자 체우기 스크린 샷 추가
- String format 으로 수정

shim506 committed:
DOCS : 리드미 줄바꿈 추가

shim506 committed:
DOCS : Intent 공부 내용 리드미 정리

shim506 committed:
DOCS : 리드미 이모지 추가

ivybae commented:
@shim506  우선 conflict을 먼저 해결해주셔야 합니다~

shim506 committed:
Merge branch 'shim506' into step2

ivybae commented on app/src/main/java/com/codesquad/kotlinphotoframe/MyImage.kt:
```diff
@@ -0,0 +1,40 @@
+            val bitmap = BitmapFactory.decodeStream(inputStream)
+            imageData = bitmap
+        }
+        return imageData as Bitmap
```
@shim506 타입캐스팅을 하신 부분에서 함께 이야기해보면 좋겠네요~
이 함수의 반환 타입이 Bitmap?이 아니고 Bitmap이 되어야 하는 이유를 한번 정리해보면 어떨까요?

shim506 replied:
이미지가 아직 로드 되지 않았다면 getImage 메서드 안에 이미지를 로드 하는 부분이 있고 경로에 대한 예외 처리를 진행했기 때문에 Bitmap 으로 타입 변환을 해보았습니다.

ivybae commented on app/src/main/java/com/codesquad/kotlinphotoframe/MyImage.kt:
```diff
@@ -0,0 +1,40 @@
+                    throw WrongPathException(filePath)
+                }
+            } catch (e: WrongPathException) {
+                Log.d(TAG, "문제 경로 : $filePath")
```
사실 이 경우에는 아래 구현이 실행되면 안되는거죠?

shim506 replied:
네 제 의도대로라만 맞습니다. 예외처리에 대한 이해가 부족했던 것 같습니다.
예외를 더 상위 에서 잡음으로서 아래 코드가 실행되지 않게 끔 수정했습니다. 
경로를 통해서 이미지를 바꾸는 행위의 최초시작점이라는 점에서 imageChangeListening() 메서드에서 예외를 catch 했습니다. 예외를 어디서 catch 할지에 대한 고민이 들었는데 혹시 말씀해주실 부분이 있을까요?

ivybae replied:
이전 코드에서는 에러가 발생해서 코드를 실행할 수 없는 상황에서도 이미지를 불러오는 요청을 하는 점이 아쉬웠습니다. 지금처럼 Exception을 발생시켜도 되고 실패한 경우에 null로 반환할 수도 있겠어요. 호출하는 시점에 예외 케이스에 대한 처리를 하신 점 좋습니다. exception을 catch 하는 시점은 지금처럼 메서드가 분리되어 있을 때는 최종적으로 처리하는 부분에서 하는 방법이 좋겠네요 👍

ivybae commented on app/src/main/java/com/codesquad/kotlinphotoframe/SecondActivity.kt:
```diff
@@ -4,55 +4,40 @@ import android.content.Intent
+        val getResult =
+            registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
+                if (it.resultCode == RESULT_OK) {
+                    Glide.with(this).load(it.data?.data).override(210, 210).centerCrop()
```
@shim506 갤러리에서 사진을 읽어올 때, Glide library를 사용하지 않고 구현할 수 있는 방법은 없을까요?

shim506 replied:
방법 3 에서 1로 수정해 보았습니다.

1. ImageView에 setImageURI() 함수를 통해 직접 추가 // 단순히 ImageView 에 사진을 추가하는데 가장 간단하게 사용할 수 있습니다.
2. MediaStore 클래스에서 getBitmap(deprecated) 을 통해 uri로 bitmap 형태로 이미지 로딩후 추가 // deprecate
3. Glide 라이브러리를 통해 추가 // Image에 대한 다양한 함수를 제공해줍니다
4. FileInputStream 으로 로드후 비트맵으로 변환하여 추가 // Image 뿐 아니라 다양한 파일을 처리할 수 있습니다. 

꼼꼼한 리뷰 감사합니다 !

ivybae replied:
여러가지 방법 파악하신 부분 좋습니다, 1, 4번 모두 학습하신 점 좋습니다 👍

ivybae reviewed (commented):
수고하셨습니다~
코멘트 확인해주세요~

shim506 committed:
FIX : 예외처리에 대한 버그 수정

-  예외에 대한 catch 를 MyImage 클래스 getImage 함수의 에서 MainActivity 의 imageChangeListening() 함수로 옮김

shim506 committed:
DOCS : Activity 학습내용 일부 추가

shim506 committed:
REFACTOR :  ImageView 에 이미지 추가하는 방식 수정

- 기존 방법 : Glide 라이브러리 사용
- 수정 방법 : ImageView 에 setImageURI 함수 사용

ivybae reviewed (approved):
수고하셨습니다, 피드백 드린 부분 모두 반영되어 머지합니다 👍
